### PR TITLE
PP718: Relay transaction with a not stored primary key

### PR DESCRIPTION
**Pull request by AndresQuijano**
## What

- Add functionality to receive a primary key and use it to sign and relay a transaction instead of an already stored account in the Account Manager.
- Add method to remove an account from the Account Manager.

## Why

- Partner requirement. 

## Refs

- [PP-718](https://rsklabs.atlassian.net/browse/PP-718)

### CODE REVIEW

File: src/AccountManager.ts [hunk 1]
```diff
@@ -40,15 +40,32 @@ export default class AccountManager {
...
+  removeAccount(addressToRemove: string) {
+    const indexToRemove = this._accounts.findIndex(function (wallet) {
+      return wallet.address === addressToRemove;
+    });
 
-    const provider = getProvider();
+    if (indexToRemove !== -1) {
+      this._accounts.splice(indexToRemove, 1);
+    } else {
+      throw new Error('Account not founded');
```
Comment (antomor) [inline src/AccountManager.ts:51]:
```suggestion
      throw new Error('Account not found');
```

File: src/RelayClient.ts [hunk 2]
```diff
@@ -364,10 +370,14 @@ class RelayClient extends EnvelopingEventEmitter {
...
-    envelopingRequest: UserDefinedEnvelopingRequest
+    envelopingRequest: UserDefinedEnvelopingRequest,
+    options?: RelayTxOptions
   ): Promise<Transaction> {
+    const signerWallet = options?.signerWallet;
+
     const { envelopingTx, activeRelay } = await this._getHubEnvelopingTx(
-      envelopingRequest
+      envelopingRequest,
+      signerWallet
```
Comment (antomor) [inline src/RelayClient.ts:380]: NIT: I wouldn't use an additional variable to store it.
```suggestion
      envelopingRequest,
      options?.signerWallet
```

File: src/RelayClient.ts [hunk 3]
```diff
@@ -392,14 +402,17 @@ class RelayClient extends EnvelopingEventEmitter {
...
   ): Promise<RelayEstimation> {
+    const signerWallet = options?.signerWallet;
+
     const {
       envelopingTx,
       activeRelay: {
         managerData: { url },
       },
-    } = await this._getHubEnvelopingTx(envelopingRequest);
+    } = await this._getHubEnvelopingTx(envelopingRequest, signerWallet);
```
Comment (antomor) [inline src/RelayClient.ts:415]: NIT: I wouldn't use an additional variable to store it.
```suggestion
    } = await this._getHubEnvelopingTx(envelopingRequest, options?.signerWallet);
```

File: test/AccountManager.test.ts [hunk 4]
```diff
@@ -98,6 +98,42 @@ describe('AccountManager', function () {
...
+          fakeAccounts;
+      });
+
+      it('Should remove only account sent as parameter', function () {
+        accountManager.removeAccount(fakeAccounts[1]?.address as string);
+
+        expect(
+          accountManager.getAccounts().length === fakeAccounts.length - 1,
+          'Incorrect length'
+        );
```
Comment (antomor) [inline test/AccountManager.test.ts:117]: I know it's the same, but for readabiltiy I'd prefer to have something like
```suggestion
        expect(
          accountManager.getAccounts()
        ).to.be.equal(fakeAccounts.length - 1);

File: test/AccountManager.test.ts [hunk 5]
```diff
@@ -98,6 +98,42 @@ describe('AccountManager', function () {
...
+        expect(
+          accountManager.getAccounts()[0] === fakeAccounts[0]?.address,
+          'First element missing'
+        );
+        expect(
+          accountManager.getAccounts()[1] === fakeAccounts[2]?.address,
+          'Second element missing'
+        );
```
Comment (antomor) [inline test/AccountManager.test.ts:125]: Actually I think that the three of the them could be replaced by
```
const expected = [fakeAccounts[0], fakeAccounts[2]];
expect(accountManager.getAccounts()).to.be.deep.equal(expected)

File: test/AccountManager.test.ts [hunk 6]
```diff
@@ -98,6 +98,42 @@ describe('AccountManager', function () {
       });
     });
 
+    describe('removeAccount()', function () {
+      let fakeAccounts: Array<Wallet>;
```
Comment (antomor) [inline test/AccountManager.test.ts:102]: Actually they're real accounts
```suggestion
      let randomAccounts: Array<Wallet>;
```

File: test/AccountManager.test.ts [hunk 7]
```diff
@@ -98,6 +98,42 @@ describe('AccountManager', function () {
...
+
+      beforeEach(function () {
+        fakeAccounts = new Array<Wallet>(3).fill(Wallet.createRandom());
```
Comment (antomor) [inline test/AccountManager.test.ts:105]: NIT: I'd store 3 in a constant `initialLength` to be used elesewhere also.

File: test/AccountManager.test.ts [hunk 8]
```diff
@@ -98,6 +98,42 @@ describe('AccountManager', function () {
...
+          'First element missing'
+        );
+        expect(
+          accountManager.getAccounts()[1] === fakeAccounts[2]?.address,
+          'Second element missing'
+        );
+      });
+
+      it('Should throw an error if the account sent as parameter does not exists', function () {
+        const inexistentWallet = Wallet.createRandom();
```
Comment (antomor) [inline test/AccountManager.test.ts:129]:
```suggestion
        const nonexistentWallet = Wallet.createRandom();
```

**Review by antomor — CHANGES_REQUESTED** [review]
@AndresQuijano thank you for the changes. I've left some comments in the tests, but overall the changes look good to me

File: src/AccountManager.ts (hunk 1 above)
Comment (AndresQuijano) [reply src/AccountManager.ts:51]: Done.

File: src/RelayClient.ts (hunk 3 above)
Comment (AndresQuijano) [reply src/RelayClient.ts:415]: Done.

File: test/AccountManager.test.ts (hunk 4 above)
Comment (AndresQuijano) [reply test/AccountManager.test.ts:117]: Done.

File: test/AccountManager.test.ts (hunk 6 above)
Comment (AndresQuijano) [reply test/AccountManager.test.ts:102]: Done.

File: test/AccountManager.test.ts (hunk 7 above)
Comment (AndresQuijano) [reply test/AccountManager.test.ts:105]: Done.

File: test/AccountManager.test.ts (hunk 8 above)
Comment (AndresQuijano) [reply test/AccountManager.test.ts:129]: Done.

File: test/AccountManager.test.ts [hunk 9]
```diff
@@ -99,38 +99,37 @@ describe('AccountManager', function () {
...
+      let randomAccounts: Array<Wallet>;
+      let initialLength: number;
 
       beforeEach(function () {
-        fakeAccounts = new Array<Wallet>(3).fill(Wallet.createRandom());
+        initialLength = 3;
+
+        randomAccounts = new Array<Wallet>(initialLength).fill(
+          Wallet.createRandom()
+        );
```
Comment (antomor) [inline test/AccountManager.test.ts:110]: I've just realised that this creates an array filled with the same object. if you check the address of the wallets created, they're all the same

File: test/AccountManager.test.ts [hunk 10]
```diff
@@ -99,38 +99,37 @@ describe('AccountManager', function () {
     });
 
     describe('removeAccount()', function () {
-      let fakeAccounts: Array<Wallet>;
+      let randomAccounts: Array<Wallet>;
+      let initialLength: number;
 
       beforeEach(function () {
-        fakeAccounts = new Array<Wallet>(3).fill(Wallet.createRandom());
+        initialLength = 3;
```
Comment (antomor) [inline test/AccountManager.test.ts:106]: I'd store this in a constant, no need to initialize it each time in a beforeEach

File: test/AccountManager.test.ts (hunk 9 above)
Comment (AndresQuijano) [reply test/AccountManager.test.ts]: Fixed.

File: test/AccountManager.test.ts (hunk 10 above)
Comment (AndresQuijano) [reply test/AccountManager.test.ts]: Done.

**Review by antomor — APPROVED** [review]

**Pull request merged**